pkg/kubelet/pod/mirror/manager.go

```diff
@@ -0,0 +1,372 @@
+// of the mirror pod always reflects the actual status of the static pod.
+// When a static pod gets deleted, the associated orphaned mirror pod will
+// also be removed.
+
+type Manager interface {
+	// Add and Delete a static pod, and the corresponding mirror pod will be created, recreated
+	// and deleted.
+	AddStaticPod(staticPod *api.Pod)
+	DeleteStaticPod(staticPod *api.Pod)
+
+	// Check a mirror pod, recreate mirror pod when needed
+	CheckMirrorPod(mirrorPod *api.Pod)
+
+	// Get current mirror pod by pod full name, if there is no mirror pod now (it hasn't been
+	// created or recreated), return nil.
+	GetMirrorPod(podFullName string) *api.Pod
+
+	// Get all mirror pods.
+	// TODO: (random-liu) Add this to kubelet.GetPods(), because mesos needs this.
+	GetMirrorPods() []*api.Pod
+
+	// Check whether a mirror pod is valid
+	IsValidMirrorPod(mirrorPod *api.Pod) bool
+}
+
+type basicManager struct {
+	// Protects staticPods and invalidMirrorPods.
+	lock sync.RWMutex
+	// Static pods indexed by full name
+	// TODO: (random-liu) Remove staticPods, and rely on pod manager is also feasible.
+	staticPods map[string]*api.Pod
+	// Buffer of invalid mirror pods
+	// TODO: (random-liu) Invalid mirror pods can also be passed to manager goroutine by channel,
+	// however that may be a little more complicated, because the channel may be full. For now
+	// just deal with invalid mirror pods in periodically cleanup. May use channel in the future.
+	invalidMirrorPods map[string]*api.Pod
+
+	// Protects mirrorPods.
+	// mirrorPods will ONLY BE MODIFIED IN THE MANAGER GOROUTINE, and read from other goroutines.
+	// mirrorPods can be READ WITHOUT LOCK in the manager goroutine, because no other goroutines
+	// will modify it.
+	mirrorPodsLock sync.RWMutex
+	// Mirror pods indexed by full name
+	mirrorPods map[string]*api.Pod
+
+	// The channel receiving mirror pod to be created
+	mirrorPodCreationChannel chan *api.Pod
+	// The channel receiving full name of mirror to be deleted
+	mirrorPodDeletionChannel chan string
+
+	podManager kubepod.Manager
+	// apiserver client
+	apiserverClient client.Interface
+}
+
+const syncPeriod = 5 * time.Second
+
+func NewManager(apiserverClient client.Interface, podManager kubepod.Manager) Manager {
+	m := &basicManager{
+		staticPods:        make(map[string]*api.Pod),
+		mirrorPods:        make(map[string]*api.Pod),
+		invalidMirrorPods: make(map[string]*api.Pod),
+		apiserverClient:   apiserverClient,
+		podManager:        podManager,
+		// There shouldn't be too many mirror pod changes, set buffer size to 100
+		mirrorPodCreationChannel: make(chan *api.Pod, 100),
+		mirrorPodDeletionChannel: make(chan string, 100),
+	}
+	// Don't start mirror pod manager goroutine if we don't have a client. This will
+	// happen on the master, where the kubelet is responsible for bootstrapping the
+	// pods of the master components.
+	if apiserverClient == nil {
+		return m
+	}
+
```

Comment:
Add a separate Start() function to start the goroutine. This is consistent with other components and will potentially save you troubles when writing the unit test (data race).